sources/web/datalab/static/datalab.js

```diff
@@ -277,6 +277,166 @@ function initializePage(dialog, saveFn) {
+const CELL_METADATA_COLLAPSED = 'hiddenCell';
+const CELL_METADATA_OUTPUT_COLLAPSED = 'hiddenOutput';
+const COLLAPSE_BUTTON_CLASS = 'fa-eye';
+const UNCOLLAPSE_BUTTON_CLASS = 'fa-eye-slash';
+const COLLAPSE_OUTPUT_BUTTON_CLASS = 'fa-compress';
+const UNCOLLAPSE_OUTPUT_BUTTON_CLASS = 'fa-expand';
+const CLEAR_BUTTON_CLASS = 'fa-minus-square-o';
+const RUN_BUTTON_CLASS = 'fa-play';
+const COLLAPSED_CELL_INNER_HTML = '-- Collapsed cell --';
+
+/**
+ * Collapse entire cell
+ */
+function toggleCollapseCell(cell, hide) {
+  // the cell is currently hidden if its input div is
+  let inputDiv = cell.cell_type === 'code' ? 'div.input' : 'div.inner_cell';
+  let isHidden = typeof(hide) !== 'undefined' ? !hide : cell.element.find(inputDiv)[0].style.display === 'none';
+  collapseSpan = cell.element.find('span.collapse-cell')[0];
+  if (isHidden) {
+    cell.element.find(inputDiv).show();
+    // for code cells, also show the output div
+    if (cell.cell_type == 'code')
+      cell.element.find('div.output_wrapper').show();
+    cell.element.find('div.cellPlaceholder').hide();
+    collapseSpan.classList.add(UNCOLLAPSE_BUTTON_CLASS);
+    collapseSpan.classList.remove(COLLAPSE_BUTTON_CLASS);
+  } else {
+    cell.element.find(inputDiv).hide();
+    // for code cells, also hide the output div
+    if (cell.cell_type == 'code')
+      cell.element.find('div.output_wrapper').hide();
+    cell.element.find('div.cellPlaceholder').show();
+    collapseSpan.classList.remove(UNCOLLAPSE_BUTTON_CLASS);
+    collapseSpan.classList.add(COLLAPSE_BUTTON_CLASS);
+  }
+  cell.metadata[CELL_METADATA_COLLAPSED] = !isHidden;
+}
+
+/**
+ * Collapse output of a cell
+ */
+function toggleCollapseCellOutput(cell, hide) {
+  // the cell's output is currently collapsed if its output div is hidden
+  let isHidden = typeof(hide) !== 'undefined' ? !hide : cell.element[0].getElementsByClassName('output')[0].style.display === 'none';
+  collapseSpan = cell.element.find('span.collapse-output')[0];
+  if (isHidden) {
+    cell.expand_output();
+    collapseSpan.classList.add(COLLAPSE_OUTPUT_BUTTON_CLASS);
+    collapseSpan.classList.remove(UNCOLLAPSE_OUTPUT_BUTTON_CLASS);
+  } else {
+    cell.collapse_output();
+    collapseSpan.classList.remove(COLLAPSE_OUTPUT_BUTTON_CLASS);
+    collapseSpan.classList.add(UNCOLLAPSE_OUTPUT_BUTTON_CLASS);
+  }
+  cell.metadata[CELL_METADATA_OUTPUT_COLLAPSED] = !isHidden;
+}
+
+/**
+ * Create an HTML button for the cell minitoolbar and return it
+ */
+function createCellMiniToolbarButton(classNames, title, callback) {
+  let buttonDiv = document.createElement('button');
+  buttonDiv.className = 'btn btn-default';
+  buttonDiv.innerHTML = '<span class="' + classNames + '"></span>';
+  buttonDiv.title = title;
+  buttonDiv.addEventListener('click', callback);
+  return buttonDiv;
+}
+
+/**
+ * Patch the cell's element to add the hoverable minitoolbar
+ */
+function addCellMiniToolbar() {
+  Jupyter.notebook.get_cells().forEach(function(cell) {
```

Comment:
I don't think that's the best spot for it. I need to be sure all the cells have rendered before I go about modifying them, that's why I'm relying on the event, does initializeNotebookApplication() guarantee this?